File: research/marker-hit-area-investigation/android_native_app/app/src/main/kotlin/com/example/markerhit/nativeapp/logging/HitLog.kt
package com.example.markerhit.nativeapp.logging

import android.util.Log
import org.json.JSONObject
import java.time.Instant

/**
 * 計測ログ単一行 JSON を logcat に "HITLOG" タグ前置で書く。
 * 受信側は `adb logcat -s HitLog:I` でも `adb logcat | grep HITLOG` でも拾える。
 */
object HitLog {
    private const val TAG = "HitLog"
    private const val PREFIX = "HITLOG"

    private fun emit(event: Map<String, Any?>) {
        val json = JSONObject(event)
        Log.i(TAG, "$PREFIX $json")
    }

    fun indexStart(app: String) = emit(
        mapOf("event" to "index_start", "app" to app, "ts" to Instant.now().toString())
    )

    fun index(
        id: String,
        shape: String,
        ratio: String,
        anchor: String,
        bitmapPxW: Int,
        bitmapPxH: Int,
        logicalPtW: Double,
        logicalPtH: Double,
        anchorScreenX: Int,
        anchorScreenY: Int,
    ) = emit(
        mapOf(
            "event" to "index",
            "id" to id,
            "shape" to shape,
            "ratio" to ratio,
            "anchor" to anchor,
            "bitmap_px_w" to bitmapPxW,
            "bitmap_px_h" to bitmapPxH,
            "logical_pt_w" to logicalPtW,
            "logical_pt_h" to logicalPtH,
            "anchor_screen_x" to anchorScreenX,
            "anchor_screen_y" to anchorScreenY,
        )
    )

    fun indexDone() = emit(
        mapOf("event" to "index_done", "ts" to Instant.now().toString())
    )

    fun markerTap(id: String) = emit(
        mapOf("event" to "tap", "id" to id, "ts" to Instant.now().toString())
    )
}

File: research/marker-hit-area-investigation/flutter_app/lib/logging/hit_log.dart
import 'dart:convert';
import 'dart:developer' as developer;

/// 計測ログを単一行 JSON で logcat に出力する。タグは "HITLOG"。
/// 受信側スクリプトは `adb logcat -s flutter:I` または `adb logcat | grep HITLOG`
/// で拾う。
class HitLog {
  static const _prefix = 'HITLOG';

  static void emit(Map<String, dynamic> event) {
    final line = '$_prefix ${jsonEncode(event)}';
    developer.log(line, name: 'flutter');
    // dart:developer log の name=flutter は logcat の flutter タグに出る。
    // 念のため print も併用すると遅延なく拾える。
    // ignore: avoid_print
    print(line);
  }

  static void indexStart({required String app}) =>
      emit({'event': 'index_start', 'app': app, 'ts': DateTime.now().toIso8601String()});

  static void index({
    required String id,
    required String shape,
    required String ratio,
    required String anchor,
    required int bitmapPxW,
    required int bitmapPxH,
    required double logicalPtW,
    required double logicalPtH,
    required int anchorScreenX,
    required int anchorScreenY,
  }) =>
      emit({
        'event': 'index',
        'id': id,
        'shape': shape,
        'ratio': ratio,
        'anchor': anchor,
        'bitmap_px_w': bitmapPxW,
        'bitmap_px_h': bitmapPxH,
        'logical_pt_w': logicalPtW,
        'logical_pt_h': logicalPtH,
        'anchor_screen_x': anchorScreenX,
        'anchor_screen_y': anchorScreenY,
      });

  static void indexDone() =>
      emit({'event': 'index_done', 'ts': DateTime.now().toIso8601String()});

  static void markerTap(String id) =>
      emit({'event': 'tap', 'id': id, 'ts': DateTime.now().toIso8601String()});
}
